

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Quick Start &mdash; Viola 1.0.0.dev10 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Signature Analysis Tutorial" href="signature_analysis.html" />
    <link rel="prev" title="Welcom to Viola Package!" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Viola
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0.dev10
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick Start</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#import-viola-to-your-script">Import Viola to your script</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-vcf-object-from-vcf-file">Create Vcf object from VCF file.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#viewing-the-contents-of-viola-vcf-object">Viewing the contents of viola.Vcf object</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#simply-print-the-object">1) Simply print the object.</a></li>
<li class="toctree-l3"><a class="reference internal" href="#generate-bedpe-like-pandas-dataframe">2) Generate BEDPE like pandas.DataFrame.</a></li>
<li class="toctree-l3"><a class="reference internal" href="#use-the-get-table-method-to-get-individual-tables-composing-the-vcf-object">3) Use the <span class="xref std std-doc">get_table()</span> method to get individual tables composing the Vcf object.</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#export-as-vcf-bedpe-file">Export as VCF/BEDPE file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#filter-vcf-object">Filter Vcf object</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#filter-with-queries-using-filter-method">1) Filter with queries using filter method</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filter-with-sv-id-using-filter-by-id-method">2) Filter with SV ID using filter_by_id method</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="signature_analysis.html">Signature Analysis Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="userguide/index.html">User Guide</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Viola</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Quick Start</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/quickstart.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="quick-start">
<span id="quickstart"></span><h1>Quick Start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h1>
<p>This is a short example how to manage VCF files.</p>
<p>After you have read this page, you can</p>
<ul class="simple">
<li><p>Import Viola package to your script.</p></li>
<li><p>Read VCF files to create viola.Vcf object.</p></li>
<li><p>Convert VCF files into BEDPE files, with any features you wish, such as INFO fields and FORMAT fields.</p></li>
<li><p>Filter SV records by any features, including genomic positions, INFO fields, FILTER fields, and FORMAT fields.</p></li>
</ul>
<p>See Also:</p>
<ul class="simple">
<li><p><a class="reference internal" href="signature_analysis.html#signature-analysis"><span class="std std-ref">Signature Analysis Tutorial</span></a></p></li>
<li><p><a class="reference internal" href="reference/index.html#api-reference"><span class="std std-ref">API Reference</span></a></p></li>
</ul>
<p>In this tutorial, we will only deal with the viola.Vcf class, but many of the features available in this class are also available in the Bedpe, MultiBedpe and MultiVcf classes.</p>
<p>Individual tutorials for classes not covered here are currently under development.</p>
<section id="import-viola-to-your-script">
<h2>Import Viola to your script<a class="headerlink" href="#import-viola-to-your-script" title="Permalink to this headline">¶</a></h2>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span> <span class="nn">viola</span>
</pre></div>
</div>
</section>
<section id="create-vcf-object-from-vcf-file">
<h2>Create Vcf object from VCF file.<a class="headerlink" href="#create-vcf-object-from-vcf-file" title="Permalink to this headline">¶</a></h2>
<p>In order to use Viola’s features, you must first run <a class="reference internal" href="reference/api/viola.read_vcf.html"><span class="doc">read_vcf</span></a> to create an <a class="reference internal" href="reference/vcf.html#vcf"><span class="std std-ref">viola.Vcf</span></a> object.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [2]: </span><span class="n">url</span> <span class="o">=</span> <span class="s1">&#39;https://raw.githubusercontent.com/dermasugita/ViolaDocs/main/docs/html/_static/tutorial.manta.vcf&#39;</span>

<span class="gp">In [3]: </span><span class="n">vcf</span> <span class="o">=</span> <span class="n">viola</span><span class="o">.</span><span class="n">read_vcf</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">variant_caller</span><span class="o">=</span><span class="s1">&#39;manta&#39;</span><span class="p">)</span> <span class="c1"># filepath, url, and file-like object are acceptable.</span>
</pre></div>
</div>
<p>Now you’re ready to perform a number of functions that the Viola package has.</p>
<p>To learn detail data structure of viola.Vcf object, see <a class="reference internal" href="userguide/data_structure.html#data-structure"><span class="std std-ref">Viola’s Data Structure</span></a>.</p>
</section>
<section id="viewing-the-contents-of-viola-vcf-object">
<h2>Viewing the contents of viola.Vcf object<a class="headerlink" href="#viewing-the-contents-of-viola-vcf-object" title="Permalink to this headline">¶</a></h2>
<p>You may want to view the content of Vcf objects after creating them.
To do so, you have three options.</p>
<section id="simply-print-the-object">
<h3>1) Simply print the object.<a class="headerlink" href="#simply-print-the-object" title="Permalink to this headline">¶</a></h3>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="go"># The column name &#39;be&#39; stands for &#39;breakend&#39;.</span>
<span class="gp">In [4]: </span><span class="nb">print</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span>
<span class="go">INFO=imprecise,svtype,svlen,end,cipos,ciend,cigar,mateid,event,homlen,homseq,svinslen,svinsseq,left_svinsseq,right_svinsseq,contig,bnd_depth,mate_bnd_depth,somatic,somaticscore,junction_somaticscore,inv3,inv5</span>
<span class="go">Documentation of Vcf object ==&gt; https://dermasugita.github.io/ViolaDocs/docs/html/reference/vcf.html</span>
<span class="go">        id              be1              be2 strand  qual svtype</span>
<span class="go">0    test1    chr1:82550461    chr1:82554226     +-  None    DEL</span>
<span class="go">1    test2    chr1:22814217    chr1:92581132     --  None    INV</span>
<span class="go">2    test3    chr1:60567906    chr1:60675941     +-  None    DEL</span>
<span class="go">3    test4    chr1:69583190    chr1:69590948     +-  None    DEL</span>
<span class="go">4    test5  chr11:104534877  chr11:104536574     +-  None    DEL</span>
<span class="go">5  test6_1  chr11:111134697   chr17:26470495     +-  None    BND</span>
<span class="go">6  test6_2   chr17:26470495  chr11:111134697     -+  None    BND</span>
</pre></div>
</div>
</section>
<section id="generate-bedpe-like-pandas-dataframe">
<h3>2) Generate BEDPE like pandas.DataFrame.<a class="headerlink" href="#generate-bedpe-like-pandas-dataframe" title="Permalink to this headline">¶</a></h3>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [5]: </span><span class="n">vcf_bedpe_like</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">to_bedpe_like</span><span class="p">()</span>

<span class="gp">In [6]: </span><span class="nb">print</span><span class="p">(</span><span class="n">vcf_bedpe_like</span><span class="p">)</span>
<span class="go">  chrom1     start1       end1 chrom2     start2       end2     name score strand1 strand2</span>
<span class="go">0   chr1   82550460   82550461   chr1   82554225   82554226    test1  None       +       -</span>
<span class="go">1   chr1   22814216   22814217   chr1   92581131   92581132    test2  None       -       -</span>
<span class="go">2   chr1   60567905   60567906   chr1   60675940   60675941    test3  None       +       -</span>
<span class="go">3   chr1   69583189   69583190   chr1   69590947   69590948    test4  None       +       -</span>
<span class="go">4  chr11  104534876  104534877  chr11  104536573  104536574    test5  None       +       -</span>
<span class="go">5  chr11  111134696  111134697  chr17   26470494   26470495  test6_1  None       +       -</span>
<span class="go">6  chr17   26470494   26470495  chr11  111134696  111134697  test6_2  None       -       +</span>
</pre></div>
</div>
<p>The way to add INFO, FILTER, and FORMAT to this bedpe-like DataFrame is explained <a class="reference internal" href="userguide/bedpe_generation.html#bedpe-generation"><span class="std std-ref">here</span></a>.</p>
</section>
<section id="use-the-get-table-method-to-get-individual-tables-composing-the-vcf-object">
<h3>3) Use the <a class="reference internal" href="reference/api/viola.Vcf.get_table.html"><span class="doc">get_table()</span></a> method to get individual tables composing the Vcf object.<a class="headerlink" href="#use-the-get-table-method-to-get-individual-tables-composing-the-vcf-object" title="Permalink to this headline">¶</a></h3>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="go"># the returned value is pd.DataFrame</span>
<span class="gp">In [7]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">get_table</span><span class="p">(</span><span class="s1">&#39;positions&#39;</span><span class="p">)</span>
<span class="gh">Out[7]: </span><span class="go"></span>
<span class="go">        id chrom1       pos1 chrom2       pos2  ... strand2 ref                 alt  qual svtype</span>
<span class="go">0    test1   chr1   82550461   chr1   82554226  ...       -   G               &lt;DEL&gt;  None    DEL</span>
<span class="go">1    test2   chr1   22814217   chr1   92581132  ...       -   T               &lt;INV&gt;  None    INV</span>
<span class="go">2    test3   chr1   60567906   chr1   60675941  ...       -   T               &lt;DEL&gt;  None    DEL</span>
<span class="go">3    test4   chr1   69583190   chr1   69590948  ...       -   T               &lt;DEL&gt;  None    DEL</span>
<span class="go">4    test5  chr11  104534877  chr11  104536574  ...       -   C               &lt;DEL&gt;  None    DEL</span>
<span class="go">5  test6_1  chr11  111134697  chr17   26470495  ...       -   T   T[chr17:26470495[  None    BND</span>
<span class="go">6  test6_2  chr17   26470495  chr11  111134697  ...       +   T  ]chr11:111134697]T  None    BND</span>

<span class="go">[7 rows x 11 columns]</span>
</pre></div>
</div>
<p>The names of all the tables in <code class="docutils literal notranslate"><span class="pre">vcf</span></code> are stored in the <a class="reference internal" href="reference/api/viola.Vcf.table_list.html"><span class="doc">table_list</span></a> attribute:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [8]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">table_list</span>
<span class="gh">Out[8]: </span><span class="go"></span>
<span class="go">[&#39;positions&#39;,</span>
<span class="go"> &#39;filters&#39;,</span>
<span class="go"> &#39;imprecise&#39;,</span>
<span class="go"> &#39;svtype&#39;,</span>
<span class="go"> &#39;svlen&#39;,</span>
<span class="go"> &#39;end&#39;,</span>
<span class="go"> &#39;cipos&#39;,</span>
<span class="go"> &#39;ciend&#39;,</span>
<span class="go"> &#39;cigar&#39;,</span>
<span class="go"> &#39;mateid&#39;,</span>
<span class="go"> &#39;event&#39;,</span>
<span class="go"> &#39;homlen&#39;,</span>
<span class="go"> &#39;homseq&#39;,</span>
<span class="go"> &#39;svinslen&#39;,</span>
<span class="go"> &#39;svinsseq&#39;,</span>
<span class="go"> &#39;left_svinsseq&#39;,</span>
<span class="go"> &#39;right_svinsseq&#39;,</span>
<span class="go"> &#39;contig&#39;,</span>
<span class="go"> &#39;bnd_depth&#39;,</span>
<span class="go"> &#39;mate_bnd_depth&#39;,</span>
<span class="go"> &#39;somatic&#39;,</span>
<span class="go"> &#39;somaticscore&#39;,</span>
<span class="go"> &#39;junction_somaticscore&#39;,</span>
<span class="go"> &#39;inv3&#39;,</span>
<span class="go"> &#39;inv5&#39;,</span>
<span class="go"> &#39;formats&#39;,</span>
<span class="go"> &#39;contigs_meta&#39;,</span>
<span class="go"> &#39;alts_meta&#39;,</span>
<span class="go"> &#39;infos_meta&#39;,</span>
<span class="go"> &#39;formats_meta&#39;,</span>
<span class="go"> &#39;filters_meta&#39;,</span>
<span class="go"> &#39;samples_meta&#39;]</span>
</pre></div>
</div>
<p>You can get any table you want.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [9]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">get_table</span><span class="p">(</span><span class="s1">&#39;formats_meta&#39;</span><span class="p">)</span> <span class="c1"># get header information of FORMAT field</span>
<span class="gh">Out[9]: </span><span class="go"></span>
<span class="go">   id number     type                    description</span>
<span class="go">0  PR   None  Integer  Spanning paired-read suppo...</span>
<span class="go">1  SR   None  Integer  Split reads for the ref an...</span>
</pre></div>
</div>
</section>
</section>
<section id="export-as-vcf-bedpe-file">
<h2>Export as VCF/BEDPE file<a class="headerlink" href="#export-as-vcf-bedpe-file" title="Permalink to this headline">¶</a></h2>
<p>You can export VCF/BEDPE files by <code class="docutils literal notranslate"><span class="pre">to_vcf</span></code>/<code class="docutils literal notranslate"><span class="pre">to_bedpe</span></code> method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">vcf</span><span class="o">.</span><span class="n">to_vcf</span><span class="p">(</span><span class="s1">&#39;/path/to/the/output.vcf&#39;</span><span class="p">)</span>
<span class="n">vcf</span><span class="o">.</span><span class="n">to_bedpe</span><span class="p">(</span><span class="s1">&#39;/path/to/the/output.bedpe&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="filter-vcf-object">
<h2>Filter Vcf object<a class="headerlink" href="#filter-vcf-object" title="Permalink to this headline">¶</a></h2>
<p>Filtering VCF file is an essential step of bioinformatics study.
viola.Vcf object provides an intuitive way to filter SV in almost any item.</p>
<p>You have two options for filtering.</p>
<section id="filter-with-queries-using-filter-method">
<h3>1) Filter with queries using filter method<a class="headerlink" href="#filter-with-queries-using-filter-method" title="Permalink to this headline">¶</a></h3>
<p>Viola has a query system that is easy to understand.</p>
<p>First, let’s look at a couple of examples.</p>
<p><strong>a. Filter with SVTYPE of the INFO field.</strong>
<code class="docutils literal notranslate"><span class="pre">syntax:</span> <span class="pre">&quot;&lt;INFO</span> <span class="pre">name&gt;</span> <span class="pre">[&lt;value</span> <span class="pre">index&gt;]</span> <span class="pre">&lt;operator&gt;</span> <span class="pre">&lt;value&gt;&quot;</span></code></p>
<ul class="simple">
<li><p>Please do not omit spaces.</p></li>
<li><p>&lt;value indexer&gt; is optional.</p></li>
<li><p>The &lt;value indexer&gt; is a 0-origin indexer which allows you to specify which of the comma-separated INFOs, such as CIPOS, should be filtered.</p></li>
<li><p>The following syntax can also be used for other INFO.</p></li>
</ul>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="go"># filter with svtype.</span>
<span class="gp">In [10]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="s1">&#39;svtype == DEL&#39;</span><span class="p">)</span>
<span class="gh">Out[10]: </span><span class="go"></span>
<span class="go">INFO=imprecise,svtype,svlen,end,cipos,ciend,cigar,mateid,event,homlen,homseq,svinslen,svinsseq,left_svinsseq,right_svinsseq,contig,bnd_depth,mate_bnd_depth,somatic,somaticscore,junction_somaticscore,inv3,inv5</span>
<span class="go">Documentation of Vcf object ==&gt; https://dermasugita.github.io/ViolaDocs/docs/html/reference/vcf.html</span>
<span class="go">      id              be1              be2 strand  qual svtype</span>
<span class="go">0  test1    chr1:82550461    chr1:82554226     +-  None    DEL</span>
<span class="go">1  test3    chr1:60567906    chr1:60675941     +-  None    DEL</span>
<span class="go">2  test4    chr1:69583190    chr1:69590948     +-  None    DEL</span>
<span class="go">3  test5  chr11:104534877  chr11:104536574     +-  None    DEL</span>
</pre></div>
</div>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="go"># example of filtering using &lt;value index&gt;</span>
<span class="go"># The code below means that if the right value of CIPOS (e.g. the value &quot;20&quot; of CIPOS=-10,20) is</span>
<span class="go"># lower than 60, the SV record will be output.</span>
<span class="gp">In [11]: </span><span class="n">using_value_idx</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="s1">&#39;cipos 1 &lt; 60&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to_vcf_like</span><span class="p">()</span>

<span class="gp">In [12]: </span><span class="nb">print</span><span class="p">(</span><span class="n">using_value_idx</span><span class="p">)</span>
<span class="go">  chrom1      pos1     id ref    alt  ...           filter                           info format  sample1_N  sample1_T</span>
<span class="go">0   chr1  82550461  test1   G  &lt;DEL&gt;  ...  MinSomaticScore  IMPRECISE;SVTYPE=DEL;SVLEN...  PR:SR  21,0:10,0  43,4:15,3</span>
<span class="go">1   chr1  22814216  test2   T  &lt;INV&gt;  ...  MinSomaticScore  IMPRECISE;SVTYPE=INV;SVLEN...     PR       24,0       35,5</span>
<span class="go">2   chr1  60567906  test3   T  &lt;DEL&gt;  ...  MinSomaticScore  SVTYPE=DEL;SVLEN=-108034;E...     PR       23,0       44,6</span>

<span class="go">[3 rows x 11 columns]</span>

<span class="gp">In [13]: </span><span class="nb">print</span><span class="p">(</span><span class="n">using_value_idx</span><span class="p">[</span><span class="s1">&#39;info&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="go">[&#39;IMPRECISE;SVTYPE=DEL;SVLEN=-3764;END=82554225;CIPOS=-51,52;CIEND=-51,52;SOMATIC;SOMATICSCORE=10&#39;</span>
<span class="go"> &#39;IMPRECISE;SVTYPE=INV;SVLEN=69766915;END=92581131;CIPOS=-51,51;CIEND=-89,90;SOMATIC;SOMATICSCORE=11;INV5&#39;</span>
<span class="go"> &#39;SVTYPE=DEL;SVLEN=-108034;END=60675940;CIPOS=-44,44;CIEND=-38,39;SOMATIC;SOMATICSCORE=18&#39;]</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>What is &lt;value index&gt;?</strong></p>
<p>Some INFO and FORMAT entries are separated by comma to store more than one value.
For example, <code class="docutils literal notranslate"><span class="pre">CIPOS</span></code> and <code class="docutils literal notranslate"><span class="pre">CIEND</span></code> entries of INFO field always store two values:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CIPOS</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">;</span><span class="n">CIEND</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">&lt;value</span> <span class="pre">index&gt;</span></code> is assigned to such data with 0-origin manner as shown below:</p>
<a class="reference internal image-reference" href="_images/value_index.png"><img alt="_images/value_index.png" src="_images/value_index.png" style="width: 300px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p>For another example, the values in FORMAT field of Manta VCF are separated by comma.</p>
<a class="reference internal image-reference" href="_images/value_index2.png"><img alt="_images/value_index2.png" src="_images/value_index2.png" style="width: 350px;" /></a>
</div>
<p><strong>b. Filter with genomic coordinates.</strong>
<code class="docutils literal notranslate"><span class="pre">syntax:</span> <span class="pre">&quot;&lt;'be1'|'be2'&gt;</span> <span class="pre">&lt;chromosome&gt;[:[&lt;start</span> <span class="pre">position&gt;]-[&lt;end</span> <span class="pre">position&gt;]]&quot;</span></code></p>
<ul class="simple">
<li><p>‘be’ stands for ‘breakend’.</p></li>
<li><p>If you skip &lt;start position&gt; with the minus sign kept, you can get all SV record younger than &lt;end position&gt;, and vice versa if you skip &lt;end position&gt;.</p></li>
<li><p>Note that &lt;start position&gt;-&lt;end position&gt; specifies genomic coordinates with left-closed, right-open interval, that is, [&lt;start position&gt;, &lt;end position&gt;).</p></li>
</ul>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="go"># filter with genomic coordinates.</span>
<span class="gp">In [14]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="s1">&#39;be1 chr11&#39;</span><span class="p">)</span>
<span class="gh">Out[14]: </span><span class="go"></span>
<span class="go">INFO=imprecise,svtype,svlen,end,cipos,ciend,cigar,mateid,event,homlen,homseq,svinslen,svinsseq,left_svinsseq,right_svinsseq,contig,bnd_depth,mate_bnd_depth,somatic,somaticscore,junction_somaticscore,inv3,inv5</span>
<span class="go">Documentation of Vcf object ==&gt; https://dermasugita.github.io/ViolaDocs/docs/html/reference/vcf.html</span>
<span class="go">        id              be1              be2 strand  qual svtype</span>
<span class="go">0    test5  chr11:104534877  chr11:104536574     +-  None    DEL</span>
<span class="go">1  test6_1  chr11:111134697   chr17:26470495     +-  None    BND</span>

<span class="gp">In [15]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="s1">&#39;be1 !chr1&#39;</span><span class="p">)</span>
<span class="gh">Out[15]: </span><span class="go"></span>
<span class="go">INFO=imprecise,svtype,svlen,end,cipos,ciend,cigar,mateid,event,homlen,homseq,svinslen,svinsseq,left_svinsseq,right_svinsseq,contig,bnd_depth,mate_bnd_depth,somatic,somaticscore,junction_somaticscore,inv3,inv5</span>
<span class="go">Documentation of Vcf object ==&gt; https://dermasugita.github.io/ViolaDocs/docs/html/reference/vcf.html</span>
<span class="go">        id              be1              be2 strand  qual svtype</span>
<span class="go">0    test5  chr11:104534877  chr11:104536574     +-  None    DEL</span>
<span class="go">1  test6_1  chr11:111134697   chr17:26470495     +-  None    BND</span>
<span class="go">2  test6_2   chr17:26470495  chr11:111134697     -+  None    BND</span>

<span class="gp">In [16]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="s1">&#39;be2 chr1:69583189-&#39;</span><span class="p">)</span>
<span class="gh">Out[16]: </span><span class="go"></span>
<span class="go">INFO=imprecise,svtype,svlen,end,cipos,ciend,cigar,mateid,event,homlen,homseq,svinslen,svinsseq,left_svinsseq,right_svinsseq,contig,bnd_depth,mate_bnd_depth,somatic,somaticscore,junction_somaticscore,inv3,inv5</span>
<span class="go">Documentation of Vcf object ==&gt; https://dermasugita.github.io/ViolaDocs/docs/html/reference/vcf.html</span>
<span class="go">      id            be1            be2 strand  qual svtype</span>
<span class="go">0  test1  chr1:82550461  chr1:82554226     +-  None    DEL</span>
<span class="go">1  test2  chr1:22814217  chr1:92581132     --  None    INV</span>
<span class="go">2  test4  chr1:69583190  chr1:69590948     +-  None    DEL</span>
</pre></div>
</div>
<p><strong>c. Filter with FORMAT table</strong>
<code class="docutils literal notranslate"><span class="pre">syntax:</span> <span class="pre">&quot;&lt;sample</span> <span class="pre">name&gt;</span> <span class="pre">&lt;FORMAT</span> <span class="pre">name&gt;</span> <span class="pre">[&lt;FORMAT</span> <span class="pre">index&gt;]</span> <span class="pre">&lt;operator&gt;</span> <span class="pre">&lt;value&gt;</span></code></p>
<ul class="simple">
<li><p>FORMAT indexer is optional. It is not required when the FORMAT isn’t separated by commas.</p></li>
<li><p>FORMAT indexer is 0-origin. Default value is 0.</p></li>
</ul>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="go"># The meaning of this code is that if the right value of the PR of sample1_T</span>
<span class="go"># in the FORMAT field (e.g. the value &quot;3&quot; in PR:SR 6,7:8,9  1,3:5,2) is greater than 5,</span>
<span class="go"># the SV records will be returned.</span>
<span class="gp">In [17]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="s1">&#39;sample1_T PR 1 &gt; 5&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to_bedpe_like</span><span class="p">(</span><span class="n">add_formats</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gh">Out[17]: </span><span class="go"></span>
<span class="go">  chrom1     start1       end1 chrom2     start2  ...  strand2 sample1_N_PR_0 sample1_N_PR_1 sample1_T_PR_0 sample1_T_PR_1</span>
<span class="go">0   chr1   60567905   60567906   chr1   60675940  ...        -             23              0             44              6</span>
<span class="go">1   chr1   69583189   69583190   chr1   69590947  ...        -             21              0             20             12</span>
<span class="go">2  chr11  104534876  104534877  chr11  104536573  ...        -             22              0             57             14</span>

<span class="go">[3 rows x 14 columns]</span>
</pre></div>
</div>
<p><strong>c. Filter with FILTER table</strong>
<code class="docutils literal notranslate"><span class="pre">syntax:</span> <span class="pre">&quot;[!]&lt;FILTER</span> <span class="pre">name&gt;&quot;</span></code></p>
<ul class="simple">
<li><p>When “!” mark is prepended, the SV records excluding &lt;FILTER name&gt; are returned.</p></li>
</ul>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [18]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="s1">&#39;PASS&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to_bedpe_like</span><span class="p">(</span><span class="n">add_filters</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gh">Out[18]: </span><span class="go"></span>
<span class="go">  chrom1     start1       end1 chrom2     start2  ...   name score strand1 strand2  PASS</span>
<span class="go">0   chr1   69583189   69583190   chr1   69590947  ...  test4  None       +       -  True</span>
<span class="go">1  chr11  104534876  104534877  chr11  104536573  ...  test5  None       +       -  True</span>

<span class="go">[2 rows x 11 columns]</span>

<span class="gp">In [19]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="s1">&#39;!PASS&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to_bedpe_like</span><span class="p">(</span><span class="n">add_filters</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gh">Out[19]: </span><span class="go"></span>
<span class="go">  chrom1     start1       end1 chrom2     start2  ...     name score strand1 strand2 MinSomaticScore</span>
<span class="go">0   chr1   82550460   82550461   chr1   82554225  ...    test1  None       +       -            True</span>
<span class="go">1   chr1   22814216   22814217   chr1   92581131  ...    test2  None       -       -            True</span>
<span class="go">2   chr1   60567905   60567906   chr1   60675940  ...    test3  None       +       -            True</span>
<span class="go">3  chr11  111134696  111134697  chr17   26470494  ...  test6_1  None       +       -            True</span>
<span class="go">4  chr17   26470494   26470495  chr11  111134696  ...  test6_2  None       -       +            True</span>

<span class="go">[5 rows x 11 columns]</span>
</pre></div>
</div>
<p><strong>d. Query can be a list</strong></p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [20]: </span><span class="n">query2_1</span> <span class="o">=</span> <span class="s1">&#39;svlen &lt; -4000&#39;</span>

<span class="gp">In [21]: </span><span class="n">query2_2</span> <span class="o">=</span> <span class="s1">&#39;svlen &gt; -10000&#39;</span>

<span class="gp">In [22]: </span><span class="n">vcf2</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">([</span><span class="n">query2_1</span><span class="p">,</span> <span class="n">query2_2</span><span class="p">],</span> <span class="n">query_logic</span><span class="o">=</span><span class="s1">&#39;and&#39;</span><span class="p">)</span>

<span class="gp">In [23]: </span><span class="n">result2</span> <span class="o">=</span> <span class="n">vcf2</span><span class="o">.</span><span class="n">to_bedpe_like</span><span class="p">(</span><span class="n">custom_infonames</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;svtype&#39;</span><span class="p">,</span> <span class="s1">&#39;svlen&#39;</span><span class="p">])</span>

<span class="gp">In [24]: </span><span class="nb">print</span><span class="p">(</span><span class="n">result2</span><span class="p">)</span>
<span class="go">  chrom1    start1      end1 chrom2    start2  ...  score strand1 strand2 svtype_0 svlen_0</span>
<span class="go">0   chr1  69583189  69583190   chr1  69590947  ...   None       +       -      DEL   -7757</span>

<span class="go">[1 rows x 12 columns]</span>
</pre></div>
</div>
<p>You can perform set operations by passing expressions to query_logic.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [25]: </span><span class="n">query3_1</span> <span class="o">=</span> <span class="s1">&#39;svtype == DEL&#39;</span>

<span class="gp">In [26]: </span><span class="n">query3_2</span> <span class="o">=</span> <span class="s1">&#39;svtype == BND&#39;</span>

<span class="gp">In [27]: </span><span class="n">query3_3</span> <span class="o">=</span> <span class="s1">&#39;somaticscore &gt; 20&#39;</span>

<span class="go"># (query3_1 or query3_2) and (query3_3)</span>
<span class="gp">In [28]: </span><span class="n">vcf3</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">filter</span><span class="p">([</span><span class="n">query3_1</span><span class="p">,</span> <span class="n">query3_2</span><span class="p">,</span> <span class="n">query3_3</span><span class="p">],</span> <span class="n">query_logic</span><span class="o">=</span><span class="s1">&#39;(0 | 1) &amp; 2&#39;</span><span class="p">)</span>

<span class="gp">In [29]: </span><span class="nb">print</span><span class="p">(</span><span class="n">vcf3</span><span class="o">.</span><span class="n">to_bedpe_like</span><span class="p">(</span><span class="n">custom_infonames</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;svtype&#39;</span><span class="p">,</span> <span class="s1">&#39;somaticscore&#39;</span><span class="p">]))</span>
<span class="go">  chrom1     start1       end1 chrom2     start2  ...  score strand1 strand2 svtype_0 somaticscore_0</span>
<span class="go">0   chr1   69583189   69583190   chr1   69590947  ...   None       +       -      DEL             47</span>
<span class="go">1  chr11  104534876  104534877  chr11  104536573  ...   None       +       -      DEL             38</span>
<span class="go">2  chr11  111134696  111134697  chr17   26470494  ...   None       +       -      BND             24</span>
<span class="go">3  chr17   26470494   26470495  chr11  111134696  ...   None       -       +      BND             24</span>

<span class="go">[4 rows x 12 columns]</span>
</pre></div>
</div>
</section>
<section id="filter-with-sv-id-using-filter-by-id-method">
<h3>2) Filter with SV ID using filter_by_id method<a class="headerlink" href="#filter-with-sv-id-using-filter-by-id-method" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Vcf.filter</span></code> is very useful, but in some situation, you may have to filter with much more complex criteria.
In such cases we recommend to use <a class="reference internal" href="reference/api/viola.Vcf.filter_by_id.html"><span class="doc">Vcf.filter_by_id</span></a> method.</p>
<p>Suppose you obtained the list of SV id, such as <code class="docutils literal notranslate"><span class="pre">['test2',</span> <span class="pre">'test4']</span></code>, as a result of quite complex criterion.
In this case, your code should be:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [30]: </span><span class="n">vcf</span><span class="o">.</span><span class="n">filter_by_id</span><span class="p">([</span><span class="s1">&#39;test2&#39;</span><span class="p">,</span> <span class="s1">&#39;test4&#39;</span><span class="p">])</span>
<span class="gh">Out[30]: </span><span class="go"></span>
<span class="go">INFO=imprecise,svtype,svlen,end,cipos,ciend,cigar,mateid,event,homlen,homseq,svinslen,svinsseq,left_svinsseq,right_svinsseq,contig,bnd_depth,mate_bnd_depth,somatic,somaticscore,junction_somaticscore,inv3,inv5</span>
<span class="go">Documentation of Vcf object ==&gt; https://dermasugita.github.io/ViolaDocs/docs/html/reference/vcf.html</span>
<span class="go">      id            be1            be2 strand  qual svtype</span>
<span class="go">0  test2  chr1:22814217  chr1:92581132     --  None    INV</span>
<span class="go">1  test4  chr1:69583190  chr1:69590948     +-  None    DEL</span>
</pre></div>
</div>
</section>
</section>
</section>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="signature_analysis.html" class="btn btn-neutral float-right" title="Signature Analysis Tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="index.html" class="btn btn-neutral float-left" title="Welcom to Viola Package!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Itsuki Sugita.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>