

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Signature Analysis Tutorial &mdash; Viola 1.0.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API Reference" href="reference/index.html" />
    <link rel="prev" title="Quick Start" href="quickstart.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Viola
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Signature Analysis Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pcawg-data-preparation">PCAWG data preparation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#other-files-preparation">Other files preparation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#generate-matrix">Generate Matrix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#importing-viola">1. Importing Viola</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reading-bedpe-files-into-viola-multibedpe-object">2. Reading BEDPE files into viola.MultiBedpe object</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reading-bed-bedgraph-files-for-annotation">3. Reading BED/BEDGRAPH files for annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#annotating-sv">4. Annotating SV</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-average-values-of-replication-timing">5. Get Average values of Replication Timing</a></li>
<li class="toctree-l3"><a class="reference internal" href="#classify-sv-and-generate-feature-matrix">6. Classify SV and Generate Feature Matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="#definition-file-syntax">Definition File Syntax</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#signature-extraction">Signature Extraction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#hyperparameter-tuning">Hyperparameter Tuning</a></li>
<li class="toctree-l2"><a class="reference internal" href="#what-does-this-function-do">What does this function do?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="userguide/index.html">User Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="release_notes/index.html">Release Notes</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Viola</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Signature Analysis Tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/signature_analysis.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="signature-analysis-tutorial">
<span id="signature-analysis"></span><h1>Signature Analysis Tutorial<a class="headerlink" href="#signature-analysis-tutorial" title="Permalink to this headline">¶</a></h1>
<p>This is a tutorial for SV signature analysis with Viola.</p>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<p>If you are new to the Viola package, please refer to the <a class="reference internal" href="quickstart.html#quickstart"><span class="std std-ref">Quick Start</span></a> first. <a class="reference internal" href="quickstart.html#quickstart"><span class="std std-ref">Quick Start</span></a> explains basic usage of <a class="reference internal" href="reference/api/viola.Vcf.html"><span class="doc">Vcf</span></a> class, which shares a lot of methods with <a class="reference internal" href="reference/api/viola.MultiBedpe.html"><span class="doc">MultiBedpe</span></a> class.</p>
<p>Installing “matplotlib” to your python environment is recommended.</p>
<p>In this tutorial, we use BEDPE files provided by <a class="reference external" href="https://dcc.icgc.org/releases/PCAWG/consensus_sv">ICGC Data Portal</a>.
Detailed instruction of data preparation is described below.</p>
<div class="section" id="pcawg-data-preparation">
<h3>PCAWG data preparation<a class="headerlink" href="#pcawg-data-preparation" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple">
<li><p>Create new directory named “resources” (Of cource you can name it otherwise) in your working directory and run <code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">resources</span></code>.</p></li>
<li><p>Download <code class="docutils literal notranslate"><span class="pre">final_consensus_sv_bedpe_passonly.icgc.public.tgz</span></code> and <code class="docutils literal notranslate"><span class="pre">final_consensus_sv_bedpe_passonly.tcga.public.tgz</span></code> to the “resources” directory.</p></li>
<li><p>Run <code class="docutils literal notranslate"><span class="pre">tar</span> <span class="pre">zxvf</span> <span class="pre">final_consensus_sv_bedpe_passonly.icgc.public.tgz</span></code> and <code class="docutils literal notranslate"><span class="pre">tar</span> <span class="pre">zxvf</span> <span class="pre">final_consensus_sv_bedpe_passonly.icgc.public.tgz</span></code></p></li>
<li><p>Now you’ve got the <code class="docutils literal notranslate"><span class="pre">icgc/</span></code> directory and the <code class="docutils literal notranslate"><span class="pre">tcga/</span></code> directory.</p></li>
<li><p>Create new directory named “pcawg” in the “resources” directory (This is also customizable).</p></li>
<li><p>Run <code class="docutils literal notranslate"><span class="pre">mv</span> <span class="pre">icgc/open/*.gz</span> <span class="pre">pcawg</span></code> and <code class="docutils literal notranslate"><span class="pre">mv</span> <span class="pre">tcga/open/*.gz</span> <span class="pre">pcawg</span></code></p></li>
<li><p>Run <code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">pcawg</span></code>, then run <code class="docutils literal notranslate"><span class="pre">gunzip</span> <span class="pre">*.gz</span></code></p></li>
<li><p>Now you’ve got decompressed BEDPE files in the “pcawg” directory.</p></li>
</ol>
</div>
<div class="section" id="other-files-preparation">
<h3>Other files preparation<a class="headerlink" href="#other-files-preparation" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple">
<li><p>Go back to the “resources” directory.</p></li>
<li><p>Visit <a class="reference external" href="https://github.com/dermasugita/Viola-SV/tree/master/examples/demo_sig/resources">here</a> and prepare same directories and files (You’ve already prepared pcawg files by above instruction).</p></li>
<li><p>Decompress <code class="docutils literal notranslate"><span class="pre">replication_timing.bedgraph.gz</span></code> by running <code class="docutils literal notranslate"><span class="pre">gunzip</span> <span class="pre">replication_timing.bedpe.gz</span></code></p></li>
<li><p>Go back to your working directory and create new Jupyter Notebook file (You can also use the <a class="reference external" href="https://github.com/dermasugita/Viola-SV/tree/master/examples/demo_sig">ipynb file</a> uploaded on GitHub.).</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Example <a class="reference external" href="https://github.com/dermasugita/Viola-SV/tree/master/examples/demo_sig">ipynb file</a> is available on GitHub!</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Due to the very large amount of data in PCAWG, the code described below will also take long time (around 50 min on Google Colab). If you want to try it in a shorter time, reduce the number of files in the directory.</p>
</div>
</div>
</div>
<div class="section" id="generate-matrix">
<h2>Generate Matrix<a class="headerlink" href="#generate-matrix" title="Permalink to this headline">¶</a></h2>
<p>Here, we explain how to generate SV feature matrix with custom SV class.
In this tutorial, we are going to classify SV by:</p>
<ul class="simple">
<li><p>Classify the SVs into DEL, DUP, INV, and TRA, and then subclassify them according to their size.</p></li>
<li><p>Common fragile sites</p></li>
<li><p>Replication timing</p></li>
</ul>
<p>This can be achieved with only eight lines of code.
The actual code is shown below.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">viola</span>
<span class="n">pcawg_bedpe</span><span class="o">=</span><span class="n">viola</span><span class="o">.</span><span class="n">read_bedpe_multi</span><span class="p">(</span><span class="s1">&#39;./resources/pcawg/&#39;</span><span class="p">,</span> <span class="n">svtype_col_name</span><span class="o">=</span><span class="s1">&#39;svclass&#39;</span> <span class="p">,</span><span class="n">exclude_empty_cases</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">bed_fragile</span> <span class="o">=</span> <span class="n">viola</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span><span class="s1">&#39;./resources/annotation/fragile_site.hg19.bed&#39;</span><span class="p">)</span>
<span class="n">bedgraph_timing</span> <span class="o">=</span> <span class="n">viola</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span><span class="s1">&#39;./resources/annotation/replication_timing.bedgraph&#39;</span><span class="p">)</span>
<span class="n">pcawg_bedpe</span><span class="o">.</span><span class="n">annotate_bed</span><span class="p">(</span><span class="n">bed</span><span class="o">=</span><span class="n">bed_fragile</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="s1">&#39;fragile&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;flag&#39;</span><span class="p">)</span>
<span class="n">pcawg_bedpe</span><span class="o">.</span><span class="n">annotate_bed</span><span class="p">(</span><span class="n">bed</span><span class="o">=</span><span class="n">bedgraph_timing</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="s1">&#39;timing&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;value&#39;</span><span class="p">)</span>
<span class="n">pcawg_bedpe</span><span class="o">.</span><span class="n">calculate_info</span><span class="p">(</span><span class="s1">&#39;($</span><span class="si">{timingleft}</span><span class="s1"> + $</span><span class="si">{timingright}</span><span class="s1">) / 2&#39;</span><span class="p">,</span> <span class="s1">&#39;timing&#39;</span><span class="p">)</span>
<span class="n">feature_matrix</span> <span class="o">=</span> <span class="n">pcawg_bedpe</span><span class="o">.</span><span class="n">classify_manual_svtype</span><span class="p">(</span><span class="n">definitions</span><span class="o">=</span><span class="s1">&#39;./resources/definitions/sv_class_definition.txt&#39;</span><span class="p">,</span> <span class="n">return_data_frame</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>Now let’s look at what each line means!</p>
<div class="section" id="importing-viola">
<h3>1. Importing Viola<a class="headerlink" href="#importing-viola" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">viola</span>
</pre></div>
</div>
</div>
<div class="section" id="reading-bedpe-files-into-viola-multibedpe-object">
<h3>2. Reading BEDPE files into viola.MultiBedpe object<a class="headerlink" href="#reading-bedpe-files-into-viola-multibedpe-object" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pcawg_bedpe</span><span class="o">=</span><span class="n">viola</span><span class="o">.</span><span class="n">read_bedpe_multi</span><span class="p">(</span><span class="s1">&#39;./resources/pcawg/&#39;</span><span class="p">,</span> <span class="n">svtype_col_name</span><span class="o">=</span><span class="s1">&#39;svclass&#39;</span><span class="p">,</span> <span class="n">exclude_empty_cases</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>This code reads all BEDPE files in the <code class="docutils literal notranslate"><span class="pre">./resources/pcawg</span></code> directory into MultiBedpe class (See <a class="reference internal" href="reference/api/viola.MultiBedpe.html"><span class="doc">MultiBedpe</span></a>).
Because the BEDPE files created by PCAWG have the <code class="docutils literal notranslate"><span class="pre">svclass</span></code> columns, we passed it to the <code class="docutils literal notranslate"><span class="pre">svtype_col_name</span></code> argument
Some BEDPE files did not have any SV records. This time we will exclude these by setting <code class="docutils literal notranslate"><span class="pre">exclude_empty_cases=True</span></code>.</p>
</div>
<div class="section" id="reading-bed-bedgraph-files-for-annotation">
<h3>3. Reading BED/BEDGRAPH files for annotation<a class="headerlink" href="#reading-bed-bedgraph-files-for-annotation" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bed_fragile</span> <span class="o">=</span> <span class="n">viola</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span><span class="s1">&#39;./resources/annotation/fragile_site.hg19.bed&#39;</span><span class="p">)</span>
<span class="n">bed_timing</span> <span class="o">=</span> <span class="n">viola</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span><span class="s1">&#39;./resources/annotation/replication_timing.bedgraph&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Reading BED and BEDGRAPH files required for custom SV classification. At the moment we do not make a clear distinction between BED files and BEDGRAPH files. This is because only the first four columns of these files are used for annotation purposes in the first place.</p>
<p><code class="docutils literal notranslate"><span class="pre">fragile_site.hg19.bed</span></code> is a BED file specifying the known common fragile site (CFS) regions.
<code class="docutils literal notranslate"><span class="pre">replication_timing.bedgraph</span></code> is a BEDGRAPH file which records the replication timing for each genome coordinate divided into bins.</p>
<p>These files were built according to the <a class="reference external" href="https://www.nature.com/articles/s41586-019-1913-9#Sec20">PCAWG paper</a>.</p>
</div>
<div class="section" id="annotating-sv">
<h3>4. Annotating SV<a class="headerlink" href="#annotating-sv" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pcawg_bedpe</span><span class="o">.</span><span class="n">annotate_bed</span><span class="p">(</span><span class="n">bed</span><span class="o">=</span><span class="n">bed_fragile</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="s1">&#39;fragile&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;flag&#39;</span><span class="p">)</span>
<span class="n">pcawg_bedpe</span><span class="o">.</span><span class="n">annotate_bed</span><span class="p">(</span><span class="n">bed</span><span class="o">=</span><span class="n">bedgraph_timing</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="s1">&#39;timing&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;value&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In this step, we annotate <code class="docutils literal notranslate"><span class="pre">pcawg_bedpe</span></code> with the Bed object we’ve just loaded.
After annotation, new INFO – ‘fragileleft’, ‘fragileright’, ‘timingleft’, and ‘timingright’ – will be added.
Because two breakends form a single SV, ‘left’ and ‘right’ suffix are added.
When <code class="docutils literal notranslate"><span class="pre">how='flag'</span></code>, annotate True/False according wether each breakend is in the range in the Bed (4th column of the Bed is ignored).
When <code class="docutils literal notranslate"><span class="pre">how='value'</span></code>, annotate the value of 4th column of Bed if the breakends hit.</p>
</div>
<div class="section" id="get-average-values-of-replication-timing">
<h3>5. Get Average values of Replication Timing<a class="headerlink" href="#get-average-values-of-replication-timing" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pcawg_bedpe</span><span class="o">.</span><span class="n">calculate_info</span><span class="p">(</span><span class="s1">&#39;($</span><span class="si">{timingleft}</span><span class="s1"> + $</span><span class="si">{timingright}</span><span class="s1">) / 2&#39;</span><span class="p">,</span> <span class="s1">&#39;timing&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>To get representative values of replication timing for each SV breakpoints, we decided to take mean values of two breakends.
This code adds new INFO named ‘timing’ by calculating mean values of ‘timingleft’ and ‘timingright’.</p>
</div>
<div class="section" id="classify-sv-and-generate-feature-matrix">
<h3>6. Classify SV and Generate Feature Matrix<a class="headerlink" href="#classify-sv-and-generate-feature-matrix" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">feature_matrix</span> <span class="o">=</span> <span class="n">pcawg_bedpe</span><span class="o">.</span><span class="n">classify_manual_svtype</span><span class="p">(</span><span class="n">definitions</span><span class="o">=</span><span class="s1">&#39;./resources/definitions/sv_class_definition.txt&#39;</span><span class="p">,</span> <span class="n">return_data_frame</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>Finally, we classified SV according to its type, size, fragile site, and replication timing. Classification criteria are written in <code class="docutils literal notranslate"><span class="pre">sv_class_definition.txt</span></code>. Syntax of this file is explained below.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">return_data_frame=True</span></code>, counts of each custom SV class for each patients are returned as pandas.DataFrame.</p>
<p>Now we successfully obtained feature matrix with custom SV classification!</p>
</div>
<div class="section" id="definition-file-syntax">
<h3>Definition File Syntax<a class="headerlink" href="#definition-file-syntax" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">name</span> <span class="s1">&#39;At fragile site DEL&#39;</span>
<span class="mi">0</span> <span class="n">fragileleft</span> <span class="o">==</span> <span class="kc">True</span>
<span class="mi">1</span> <span class="n">fragileright</span> <span class="o">==</span> <span class="kc">True</span>
<span class="mi">2</span> <span class="n">svtype</span> <span class="o">==</span> <span class="n">DEL</span>
<span class="n">logic</span> <span class="p">(</span><span class="mi">0</span> <span class="o">|</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="mi">2</span>

<span class="n">name</span> <span class="s1">&#39;At fragile site DUP&#39;</span>
<span class="mi">0</span> <span class="n">fragileleft</span> <span class="o">==</span> <span class="kc">True</span>
<span class="mi">1</span> <span class="n">fragileright</span> <span class="o">==</span> <span class="kc">True</span>
<span class="mi">2</span> <span class="n">svtype</span> <span class="o">==</span> <span class="n">DUP</span>
<span class="n">logic</span> <span class="p">(</span><span class="mi">0</span> <span class="o">|</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="mi">2</span>

<span class="n">name</span> <span class="s1">&#39;&lt;50 kb early DEL&#39;</span>
<span class="mi">0</span> <span class="n">svlen</span> <span class="o">&gt;</span> <span class="o">-</span><span class="mi">50000</span>
<span class="mi">1</span> <span class="n">timing</span> <span class="o">&gt;</span> <span class="mf">66.65</span>
<span class="mi">2</span> <span class="n">svtype</span> <span class="o">==</span> <span class="n">DEL</span>
<span class="n">logic</span> <span class="mi">0</span> <span class="o">&amp;</span> <span class="mi">1</span> <span class="o">&amp;</span> <span class="mi">2</span>

<span class="n">name</span> <span class="s1">&#39;&lt;50 kb mid DEL&#39;</span>
<span class="mi">0</span> <span class="n">svlen</span> <span class="o">&gt;</span> <span class="o">-</span><span class="mi">50000</span>
<span class="mi">1</span> <span class="n">timing</span> <span class="o">&gt;</span> <span class="mf">33.35</span>
<span class="mi">2</span> <span class="n">svtype</span> <span class="o">==</span> <span class="n">DEL</span>
<span class="n">logic</span> <span class="mi">0</span> <span class="o">&amp;</span> <span class="mi">1</span> <span class="o">&amp;</span> <span class="mi">2</span>

<span class="n">name</span> <span class="s1">&#39;&lt;50 kb late DEL&#39;</span>
<span class="mi">0</span> <span class="n">svlen</span> <span class="o">&gt;</span> <span class="o">-</span><span class="mi">50000</span>
<span class="mi">1</span> <span class="n">svtype</span> <span class="o">==</span> <span class="n">DEL</span>
<span class="n">logic</span> <span class="mi">0</span> <span class="o">&amp;</span> <span class="mi">1</span>
</pre></div>
</div>
<p>This is an example of definition file of custom SV classification.</p>
<p>Each SV class is defined by a syntax like the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">name</span> <span class="s1">&#39;&lt;SV class name&gt;&#39;</span>
<span class="mi">0</span> <span class="o">&lt;</span><span class="n">condition</span><span class="o">&gt;</span>
<span class="mi">1</span> <span class="o">&lt;</span><span class="n">condition</span><span class="o">&gt;</span>
<span class="mi">2</span> <span class="o">&lt;</span><span class="n">condition</span><span class="o">&gt;</span>
<span class="o">...</span>
<span class="n">logic</span> <span class="o">&lt;</span><span class="nb">set</span> <span class="n">operation</span><span class="o">&gt;</span>
</pre></div>
</div>
<ul class="simple">
<li><p>The syntax of &lt;condition&gt; is the same as query passed to Vcf.filter method (See <a class="reference internal" href="quickstart.html#quickstart"><span class="std std-ref">Quick Start</span></a>).</p></li>
<li><p>The numbers written in the left of each &lt;condition&gt; can be omitted.</p></li>
<li><p>Use numbers correspoinding to each &lt;condition&gt; for the &lt;set operation&gt;</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The order of the SV class definition is very important. The <code class="docutils literal notranslate"><span class="pre">classify_manual_svtype</span></code> method reads the definition file in order from the top, so that the SV class definitions written higher up in the file take precedence. Thus, in above example, Deletions that both satisfy ‘At fragile site DEL’ and ‘&lt;50 kb early DEL’ criteria, they are classified as ‘At fragile site DEL’, not ‘&lt;50 kb early DEL’.</p>
</div>
</div>
</div>
<div class="section" id="signature-extraction">
<h2>Signature Extraction<a class="headerlink" href="#signature-extraction" title="Permalink to this headline">¶</a></h2>
<p>Viola offers the function, <a class="reference internal" href="reference/api/viola.SV_signature_extractor.html"><span class="doc">viola.SV_signature_extractor</span></a> which performs non-negative matrix factorization (NMF) and cluster stability evaluation at the same time.</p>
<p>Before diving into the details of <a class="reference internal" href="reference/api/viola.SV_signature_extractor.html"><span class="doc">viola.SV_signature_extractor</span></a>, let’s look at an example usage.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">feature_matrix</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;others&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># remove &quot;others&quot; column</span>

<span class="n">result_silhouette</span><span class="p">,</span> <span class="n">result_metrics</span><span class="p">,</span> <span class="n">exposure_matrix</span><span class="p">,</span> <span class="n">signature_matrix</span> <span class="o">=</span> <span class="n">viola</span><span class="o">.</span><span class="n">SV_signature_extractor</span><span class="p">(</span>
<span class="n">feature_matrix</span><span class="p">,</span> <span class="n">n_iter</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;testRun&#39;</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;nndsvda&#39;</span><span class="p">,</span> <span class="n">solver</span><span class="o">=</span><span class="s1">&#39;mu&#39;</span><span class="p">,</span> <span class="n">beta_loss</span><span class="o">=</span><span class="s1">&#39;kullback-leibler&#39;</span><span class="p">,</span> <span class="n">max_iter</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1</span>
<span class="p">)</span>

<span class="c1">######## STDOUT ########</span>
<span class="c1"># testRun: finished NMF</span>
<span class="c1"># testRun: finished kmeans clustering</span>
<span class="c1"># testRun: finished all steps</span>
<span class="c1"># Silhouette Score: 0.9997561882187715, kullback-leibler: 98901.64510245458</span>
<span class="c1">#</span>
<span class="c1"># ==================</span>
</pre></div>
</div>
<p><a class="reference internal" href="reference/api/viola.SV_signature_extractor.html"><span class="doc">viola.SV_signature_extractor</span></a> outputs four returns. We will explain them one by one.</p>
<ul class="simple">
<li><dl class="simple">
<dt>result_silhouette:</dt><dd><p>This is the score that indicates the stability (reproducibility) of the SV signature. Detailed explanations are given below.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>result_metrics:</dt><dd><p>An error between the original matrix and the product of the factored matrice.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>exposure_matrix:</dt><dd><p>An exposure matrix with (n_samples × n_signatures).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>signature_matrix:</dt><dd><p>A result SV signature matrix with (n_signatures × n_features). Here, <code class="docutils literal notranslate"><span class="pre">n_features</span></code> means “the number of custom SV class”</p>
</dd>
</dl>
</li>
</ul>
<p>Viola employs <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.NMF.html">NMF class of Scikit Learn</a> for signature extraction.
So, <a class="reference internal" href="reference/api/viola.SV_signature_extractor.html"><span class="doc">viola.SV_signature_extractor</span></a> receives completely the same arguments of <a href="#id1"><span class="problematic" id="id2">`sklearn-decomposition.NMF&lt;NMF class of Scikit Learn&gt;`_</span></a> except the first three, which are the original arguments of Viola.</p>
<p>Here are the explanations for Viola original arguments of <a class="reference internal" href="reference/api/viola.SV_signature_extractor.html"><span class="doc">viola.SV_signature_extractor</span></a></p>
<ul class="simple">
<li><dl class="simple">
<dt>X:</dt><dd><p>A (n_sample, n_features) shaped ndarray or DataFrame. Each column represents each SV class.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>n_iter:</dt><dd><p>Number of bootstrap resampling from X. The higher this number, the more reliable the result.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>name:</dt><dd><p>Name of this run.</p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="hyperparameter-tuning">
<h2>Hyperparameter Tuning<a class="headerlink" href="#hyperparameter-tuning" title="Permalink to this headline">¶</a></h2>
<p>For signature analysis, hyperparameters such as the number of signatures and the loss function need to be determined.</p>
<p>Here is an example of the hyperparameter tuning.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ls_silhouette</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">ls_metrics</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">ls_exposure</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">ls_signature</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">num_signature</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">15</span><span class="p">):</span>
   <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Number of Signature: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_signature</span><span class="p">))</span>
   <span class="n">result_silhouette</span><span class="p">,</span> <span class="n">result_metrics</span><span class="p">,</span> <span class="n">exposure_matrix</span><span class="p">,</span> <span class="n">signature_matrix</span> <span class="o">=</span> <span class="n">viola</span><span class="o">.</span><span class="n">SV_signature_extractor</span><span class="p">(</span>
      <span class="n">feature_matrix</span><span class="p">,</span>
      <span class="n">n_iter</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
      <span class="n">name</span><span class="o">=</span><span class="s1">&#39;testRun&#39;</span><span class="p">,</span>
      <span class="n">n_components</span><span class="o">=</span><span class="n">num_signature</span><span class="p">,</span>
      <span class="n">init</span><span class="o">=</span><span class="s1">&#39;nndsvda&#39;</span><span class="p">,</span>
      <span class="n">solver</span><span class="o">=</span><span class="s1">&#39;mu&#39;</span><span class="p">,</span>
      <span class="n">beta_loss</span><span class="o">=</span><span class="s1">&#39;kullback-leibler&#39;</span><span class="p">,</span>
      <span class="n">max_iter</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span>
      <span class="n">random_state</span><span class="o">=</span><span class="mi">1</span>
   <span class="p">)</span>

 <span class="n">ls_silhouette</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result_silhouette</span><span class="p">)</span>
 <span class="n">ls_metrics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result_metrics</span><span class="p">)</span>
 <span class="n">ls_exposure</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exposure_matrix</span><span class="p">)</span>
 <span class="n">ls_signature</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">signature_matrix</span><span class="p">)</span>

<span class="c1">######## STDOUT ########</span>
<span class="c1"># Number of Signature: 2</span>
<span class="c1"># testRun: finished NMF</span>
<span class="c1"># testRun: finished kmeans clustering</span>
<span class="c1"># testRun: finished all steps</span>
<span class="c1"># Silhouette Score: 0.9995314504820613, kullback-leibler: 98898.80582214911</span>

<span class="c1"># ==================</span>

<span class="c1"># Number of Signature: 3</span>
<span class="c1"># testRun: finished NMF</span>
<span class="c1"># testRun: finished kmeans clustering</span>
<span class="c1"># testRun: finished all steps</span>
<span class="c1"># Silhouette Score: 0.9996664503915801, kullback-leibler: 75815.89695274398</span>

<span class="c1"># ==================</span>

<span class="c1"># Number of Signature: 4</span>
<span class="c1"># testRun: finished NMF</span>
<span class="c1"># testRun: finished kmeans clustering</span>
<span class="c1"># testRun: finished all steps</span>
<span class="c1"># Silhouette Score: 0.9219878473698799, kullback-leibler: 65871.09184405269</span>

<span class="c1"># ==================</span>

<span class="c1"># ...</span>
</pre></div>
</div>
<p>In order to determine the number of signatures based on the balance between the silhouette score and the Kullback-Leibler distance, let’s visualise the data.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">matplotlib</span> <span class="n">inline</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>

<span class="n">x</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ls_silhouette</span><span class="p">))]</span>

<span class="n">ln1</span> <span class="o">=</span> <span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">ls_silhouette</span><span class="p">,</span><span class="s1">&#39;C0&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;silhouette score&quot;</span><span class="p">)</span>
<span class="n">ax2</span> <span class="o">=</span> <span class="n">ax1</span><span class="o">.</span><span class="n">twinx</span><span class="p">()</span>
<span class="n">ln2</span> <span class="o">=</span> <span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">ls_metrics</span><span class="p">,</span><span class="s1">&#39;C1&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;KL Divergence&quot;</span><span class="p">)</span>
<span class="n">h1</span><span class="p">,</span> <span class="n">l1</span> <span class="o">=</span> <span class="n">ax1</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()</span>
<span class="n">h2</span><span class="p">,</span> <span class="n">l2</span> <span class="o">=</span> <span class="n">ax2</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">h1</span><span class="o">+</span><span class="n">h2</span><span class="p">,</span> <span class="n">l1</span><span class="o">+</span><span class="n">l2</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Silhouette Score vs KL Divergence&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_major_locator</span><span class="p">(</span><span class="n">plt</span><span class="o">.</span><span class="n">MultipleLocator</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Number of Signatures&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="_images/silhouette_vs_kld.png" src="_images/silhouette_vs_kld.png" />
<p>Now we can have a discussion that the number of signatures should be around 7-9.</p>
<p>The biological consideration is needed to determine the most appropreate number (Not the main topic here).</p>
</div>
<div class="section" id="what-does-this-function-do">
<h2>What does this function do?<a class="headerlink" href="#what-does-this-function-do" title="Permalink to this headline">¶</a></h2>
<p>Editing in progress…</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="reference/index.html" class="btn btn-neutral float-right" title="API Reference" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="quickstart.html" class="btn btn-neutral float-left" title="Quick Start" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Itsuki Sugita.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>